File: sumOfFactorization.js
function sum(array){
    return array.reduce( (a,b)=>a+b, 0);
}

/**
 * Returns factors of that number, except for 1 and number
 * @param {Integer} num 
 * @returns {Array} result
 */
function findFactors(num){
    const result = [];
    for( let i=2; i <= num ** (1/2); ++i ){
        if(num % i === 0){
            result.push(i);
            result.push(num / i);
        }
    }

    return result;
}
console.log(findFactors(879687888877588));
//for(let i = 1; i < 10000000; ++i){
    //let sum_of_factors = sum(findFactors(i));
    //if(sum_of_factors-i === 1 ||sum_of_factors-i === -1) console.log(`i: ${i}   Sum:${sum_of_factors}  Difference:${sum_of_factors-i}`);
    // if(sum(findFactors(i)) === i){
    //     console.log(i);
    // }
    //console.log(findFactors(650));
//}

File: sumOfFactorization.cpp
#include <iostream>
#include <numeric>
#include <cmath>
#include <vector>


auto sum(const std::vector<int>& factors) -> int;

auto find_factors(const int number) -> std::vector<int>;

auto main() -> int{
    for (size_t i = 2; i < 1000000000; i++)
    {   
        //lemma that shows that a multiple of 6 cannot fufil the desired property
        if(i % 6 == 0){continue;}

        const auto sum_amount = sum(find_factors(i));
        if( sum_amount == i){
            std::cout << sum(find_factors(i)) << std::endl;
        } 

        if( sum_amount - 1 == i){
            std::cout << "i:" << i << " difference: 1" << std::endl;
        }
    }
};

auto sum(const std::vector<int>& factors) -> int{  
    return std::accumulate(factors.begin(), factors.end(), 0);
}

auto find_factors(const int number) -> std::vector<int>{
    auto factors = std::vector<int>{};
    factors.reserve(std::sqrt(number) / 1);
    for (size_t i = 2; i <= (std::sqrt(number)/1); ++i)
    {
        if(number % i == 0){
            factors.emplace_back(i);
            factors.emplace_back(number/i);
        }
        
        //if(sum(factors)>= i+2){break;}
    }
    
    return factors;
}

